Basic support for themeing plots

noelwelsh opened the pull request:
This WIP adds basic support for theming plots. What we have:

* A way to represent style values that can be overridden by a theme, or override a theme
* A way to represent a theme
* Very basic implementation for `Layout`

Things to do:

- [x] Implement some actual themes
- [x] Pass themes through from `Plot`
- [x] Complete documentation
- [x] Correct typos: themeing -> theming

To consider:

- Should there be a hierarchy of themes, and inheritance between them.
- Clean up `Glyph` implementation, so it doesn't have its own theming code
- Add colour back to our examples. I.e. add a default theme that looks ok.

noelwelsh committed:
Basic sketch of theme implementation

noelwelsh committed:
Start documenting theme system

noelwelsh committed:
Wordsmithing

noelwelsh committed:
Complete basic implementation of theming

- Plots have themes, which apply to layers
- There is a very basic default theme

noelwelsh committed:
Update Cats dependency

noelwelsh committed:
Fix typos

noelwelsh committed:
Add some themes, taken from the base16 themes

noelwelsh committed:
Add basic documentation for themes

noelwelsh commented:
Closes #3

danielost reviewed (changes requested):
It works great with the scatter plot, but there is an issue with lines and curves. I think that might happen because of the default `fillColor` option.

![image](https://github.com/creativescala/chartreuse/assets/72448226/0f01e619-37e1-489c-aceb-f922148bb5ef)

noelwelsh commented:
@danielost Good catch. I've added creativescala/doodle#136 to fix the SVG backend. As a workaround I'll make `Line` and `Curve` override any theme fill with an empty fill.

Legends don't follow themes either, so that's another fix I need to make.

noelwelsh committed:
Add some utility methods

noelwelsh committed:
Improve theme support

- Add methods that are more convenient for overriding themes
- Documentation on methods
- Add some tests

noelwelsh committed:
Override fill and correctly apply theme

- Line and Curve should not have a fill, so their default theme
overrides the fillColor to `None`

- Correctly apply the theme when drawing the shape.

noelwelsh committed:
Update example to use correct methods

noelwelsh committed:
Legend picks up theme from layer

- Legend uses the themed color from each layer when drawing
- All layout's have themeable values
- Scatter is updated to be themeable
- Unnecessary complications in Glyph are removed

noelwelsh committed:
Fill text with Black

SVG doesn't fill text by default. This should be taken from a style, but
we don't yet have this kind of style.

noelwelsh committed:
Get rid of the ugly scatter plot example

noelwelsh committed:
Fix typo

noelwelsh committed:
Add additional themes

Themes are taken from Matplotlib

noelwelsh committed:
Improve theme documentation

- Give examples of setting theme and overriding theme
- Create a small gallery of themes
